lib/cloudant.js

```diff
@@ -524,21 +531,52 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
           model, where, data, options);
   var self = this;
   var mo = self.selectModel(model);
+  var idName = self.idName(model);
+  // if idName exists, change the idName to be the default `_id` key
+  if (idName && data[idName]) {
+    data._id = data[idName];
+    delete data[idName];
+  }
+  var res = {};
+  var docList = [];
+
   self.all(model, {where: where}, {raw: true}, function(err, docs) {
     if (err) return cb(err, docs);
     if (docs.length === 0) return cb(null, {count: 0});
 
     data = getPlainJSONData.call(self, model, data);
+
+    /**
+     * for each of the docs from the given filter, only update the instance where the revs match
+     * behavior of bulk update in cloudant:
+       - if both _id and _rev are not provided, return empty list
+       - if _id is provided and _rev is not provided, throw error
+       - if _id is not provided and _rev is provided, do update
+       - if both _id and _rev are provided, do update
+    **/
     async.each(docs, function(doc, cb) {
-      _.mergeWith(doc, data, function(dest, src) { return src; });
+      if (doc._id === data._id || doc._rev === data._rev) {
+        delete doc._id;
+        delete doc._rev;
+        _.mergeWith(doc, data, function(dest, src) { return src; });
```

Comment:
Given the comment above, are we certain that `_.mergeWith` will be merging the correct items in each array together, or does it only line them up by index?

You might want a test to validate that the correct objects are being combined within the arrays _in the correct way_ (eg. the `docs` array's version of `{id: 1}` is being merged with the `data` array's version of `{id: 1}`)